### PR TITLE
Bushwack_Bot V1.2

### PR DESCRIPTION
**Pull request by Fletcha**
Submission for Bushwack bot:- Created by `IceCube`

### CODE REVIEW

File: bots/bushwack/bushwack_main.lua [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+		skills.abilSplit = unitSelf:GetAbility(2)
+		skills.abilToxin = unitSelf:GetAbility(3)
+		skills.attributeBoost = unitSelf:GetAbility(4)
+	end
+	
+	if unitSelf:GetAbilityPointsAvailable() <= 0 then
+		return
+	end
+	
+	if skills.abilToxin:CanLevelUp() then
```
Comment (mtesija) [inline bots/bushwack/bushwack_main.lua]: This should be updated to the new skill build function

File: bots/bushwack/bushwack_main.lua [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+-- bonus aggression points if a skill/item is available for use
+object.nDartUp = 10
+object.nJumpUp = 10
+object.nEnergizerUp = 8
+-- bonus aggression points that are applied to the bot upon successfully using a skill/item
+object.nDartUse = 10
+object.nJumpUse = 10
+object.nEnergizerUse = 10
+--thresholds of aggression the bot must reach to use these abilities
+object.nDartThreshold = 20
```
Comment (mtesija) [inline bots/bushwack/bushwack_main.lua]: Thresholds less than 21 are always used, these should be reworked

File: bots/bushwack/bushwack_main.lua [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+		--Jump
+	local nRange = abilJump:GetRange()
+	if abilJump:CanActivate() and nTargetDistanceSq < ((nRange * nRange) * 4) and  nLastHarassUtility > botBrain.nJumpThreshold then
+		local vecAbilityTarget = unitTarget:GetPosition()
+		bAction = core.OrderAbilityPosition(botBrain, abilJump, vecAbilityTarget)
+	end	
+		
+		--Dart
+	if not bActionTaken and abilDart:CanActivate()then
+		local nRange = abilDart:GetRange()
```
Comment (mtesija) [inline bots/bushwack/bushwack_main.lua:198]: Should check CanSee for single target abilities.

File: bots/bushwack/bushwack.bot [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<bot
+	name="BushwackBot"
+	description="bot_description_bushwack"
```
Comment (Anakonda) [inline bots/bushwack/bushwack.bot]: You didn't add the description to the stringtables, If you want to make your own see https://github.com/honteam/Heroes-of-Newerth-Bots/blob/Community/stringtables/interface_en.str#L10596 else we make one for you.

File: bots/bushwack/bushwack_main.lua [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+object.nDartUp = 10
+object.nJumpUp = 10
+object.nEnergizerUp = 8
+-- bonus aggression points that are applied to the bot upon successfully using a skill/item
+object.nDartUse = 12
+object.nJumpUse = 10
+object.nEnergizerUse = 10
+--thresholds of aggression the bot must reach to use these abilities
+object.nDartThreshold = 22
+object.nJumpThreshold = 20
```
Comment (Anakonda) [inline bots/bushwack/bushwack_main.lua]: Harass hero does not run if the utility is less than 20 so these do nothing. Increase them or remove them.

File: bots/bushwack/bushwack_main.lua [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+	local abilDart = skills.abilDart
+	local abilJump = skills.abilJump
+	local tLocalEnemyHeroes = core.localUnits["EnemyHeroes"]
+	local tLocalAllyHeroes = core.localUnits["AllyHeroes"]
+	 
+		--Jump
+	local nRange = abilJump:GetRange()
+	if abilJump:CanActivate() and nTargetDistanceSq < ((nRange * nRange) * 4) and  nLastHarassUtility > botBrain.nJumpThreshold then
+		local vecAbilityTarget = unitTarget:GetPosition()
+		bAction = core.OrderAbilityPosition(botBrain, abilJump, vecAbilityTarget)
```
Comment (Anakonda) [inline bots/bushwack/bushwack_main.lua]: Should be bActionTaken

File: bots/bushwack/bushwack_main.lua [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+		
+		--Dart
+	if not bActionTaken and abilDart:CanActivate() and core.CanSeeUnit(botBrain, unitTarget)
+		local nRange = abilDart:GetRange()
+		if ( nLastHarassUtility > botBrain.nDartThreshold and unitSelf:GetMana() > 0.80 ) or ( unitTarget:GetHealth() < 0.45 )  then
+			if nTargetDistanceSq < (nRange * nRange) then  --Dart if target is in range
+				bActionTaken = core.OrderAbilityEntity(botBrain, abilDart, unitTarget)
+			elseif abilJump:CanActivate() and nTargetDistanceSq > (nRange * nRange) then --Jump at target in range
+				local vecAbilityTarget = unitTarget:GetPosition()
+				bAction = core.OrderAbilityPosition(botBrain, abilJump, vecAbilityTarget)
```
Comment (Anakonda) [inline bots/bushwack/bushwack_main.lua]: bActionTaken

File: bots/bushwack/bushwack_main.lua [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+	return false
+end
+object.HealAtWellUtilityOld = behaviorLib.HealAtWellBehavior["Utility"]
+behaviorLib.HealAtWellBehavior["Utility"] = HealAtWellUtilityOverride
+
+----------------------------------
+--	Bushwack items
+----------------------------------
+behaviorLib.StartingItems = {"2 Item_DuckBoots", "2 Item_MinorTotem", "Item_HealthPotion", "Item_RunesOfTheBlight"}
```
Comment (Anakonda) [inline bots/bushwack/bushwack_main.lua]: Please move items at top of file - before skills.

File: bots/bushwack/bushwack_main.lua [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+	local unitSelf = self.core.unitSelf
+    if  skills.abilDart == nil then
+	skills.abilDart = core.WrapInTable(unitSelf:GetAbility(0))
+        skills.abilDart.nLastCastTime = 0
+        skills.abilJump = unitSelf:GetAbility(1)
+        skills.abilSplit = core.WrapInTable(unitSelf:GetAbility(2))
+        skills.abilSplit.nLastCastTime = 0
+        skills.abilToxin = unitSelf:GetAbility(3)
+        skills.abilAttributeBoost = unitSelf:GetAbility(4)
```
Comment (Anakonda) [inline bots/bushwack/bushwack_main.lua]: abilAttributeBoost is unused. Remove it.

File: bots/bushwack/bushwack.bot (hunk 4 above)
Comment (Fletcha) [inline bots/bushwack/bushwack.bot]: I added a description in "bots/bushwack/bushwack.bot" . Do you want me to create a new pull request for it?

File: bots/bushwack/bushwack_main.lua (hunk 9 above)
Comment (Fletcha) [inline bots/bushwack/bushwack_main.lua]: I changed everything you suggested, thanks for helping me with this. :)

**Pull request merged**